app/src/main/java/com/samco/trackandgraph/antlr/evaluation/Evaluation.kt

```diff
@@ -0,0 +1,99 @@
+import com.samco.trackandgraph.antlr.parsing.DatatransformationFunctionParserFacade
+import com.samco.trackandgraph.database.entity.DataPoint
+
+fun inferDatatype(data: List<DataPoint>) : DataType {
+    return DataType.NUMERICAL
+}
+
+class EvaluationModel {
+
+    fun run(code: String, inputFeatures: Map<String, List<DataPoint>> = emptyMap()) : Map<String, Value> {
+        val parseResult = DatatransformationFunctionParserFacade.parse(code, inputFeatures.keys)
+        when (parseResult.errors.size) {
+            0 -> {}
+            1 -> throw parseResult.errors[0]
+            else -> throw ListOfErrors(parseResult.errors)
+        }
+
+        return this.evaluate(parseResult.root!!, inputFeatures)
+    }
+
+    private fun evaluate(
+        parsedTree: DatatransformationFunction,
+        inputFeatures: Map<String, List<DataPoint>> = emptyMap<String, List<DataPoint>>()
+    ) : Map<String, Value> {
+        var context : Map<String, Value> = inputFeatures.mapValues { list -> DatapointsValue(list.value, inferDatatype(list.value)) }
+        for (statement in parsedTree.statements) {
+            context = when (statement) {
+                is VarDeclaration -> statement.evaluate(context)
+                is Assignment -> statement.evaluate(context)
+                is Print -> statement.evaluate(context)
+                else -> throw UnsupportedOperationException(this.javaClass.canonicalName)
+            }
+
+        }
+
+        return context
+    }
+}
+
+fun Expression.evaluate(context: Map<String, Value>) : Value {
+    try {
+        return when (this) {
+            is IntLit -> NumberValue(this.value.toInt())
+            is DecLit -> NumberValue(this.value.toDouble())
+            is StringLit -> StringValue(this.value)
+            is TimeperiodLit -> TimeValue(this.context)
+
+            is MultiplicationExpression -> this.left.evaluate(context) * this.right.evaluate(context)
+            is DivisionExpression       -> this.left.evaluate(context) / this.right.evaluate(context)
+            is SumExpression            -> this.left.evaluate(context) + this.right.evaluate(context)
+            is SubtractionExpression    -> this.left.evaluate(context) - this.right.evaluate(context)
+
+            is FunctionCall             -> callFunction(functionName, args.map { it.evaluate(context) })
+
+            is VarReference -> context.get(this.varName) ?: throw NotDeclaredError(this.varName, this.position!!)
+            else -> throw UnsupportedOperationException(this.javaClass.canonicalName)
+//            else -> throw UnsupportedArgumentTypeError(
+//                this.toString(), this.javaClass.canonicalName!!,
+//                "Number, Variable Reference, Arithmetic", this.position!!
+//            )
+        }
+    } catch (e: Error) {
+        e.position = this.position!!
+        throw e
+    }
+
+}
+
+
+
+fun VarDeclaration.evaluate(context: Map<String, Value>) : HashMap<String, Value> {
```

Comment:
Okay so here is rough idea for the functions:
First I never wanted to load the full data base into memory :D
So my basic assumption is that a lot of users would like (slightly) more complex data manipulation, and we should assume that they want to use code or anything like that as little as possible. So my idea would be to ship the feature with ~10+ presets (stuff like difference between points,  cumulative sum, etc.) We could even convert the existing options for aggregation/averaging to these code presets. So then we have these functions which operate mostly on one data series, so we only need to load this one from the database prior and pass it to the eval function. 

Assuming that we have presets and that 90% of users never want to change/code anything I think we should not have the functions at the same location as the graphs and track groups. Just think about how often you track, how often you want to look at graphs and then how often you want to adapt your custom weekly total into monthly median function.

So I would add a new entry on the burger menu thing just for functions where they look like track groups but are among another. Once you click on a function to edit it, each function has a "header" outside the code that specifies how many input data series it takes and assigns names to them to use in the code. It might make sense to also allow for other custom inputs per function e.g. the aggregation window so we can reuse the code for day/week/month etc. But that would be slightly harder to solve UI wise I think.

Another thing would be to have applied functions which can look like normal track-groups but you can only view the data inside (e.g. if I would want to look at my weekly totals of something). These would fit right with the ui elements we already have.

Reply:
Ok I like your idea having a separate input for the features we want to reference. I guess we can do it this way. I still wonder if it isn't worth using an interface to access the data as we need it though rather than reading it all into memory. I'm not sure. Maybe we can address this later.

I also like your idea of having something like a read only tracker to represent the function. I think we should combine our ideas here. From a group you click the + in the top right and you see the drop down list ["Tracker", "Graph or Statistic", "Group", "Function"] (maybe we find a more user friendly name than function? idk) .. If you select function it takes you to a setup activity that is essentially your code editor and feature inputs UI. Once you click accept you come back to the group and now in your list for that group you see your function and it looks a bit like a tracker. If you click on it you see the history of data points that it produces and it contains much of the same data on the card as a regular tracker but there's no + button so you can't add data to it directly. I will make sure it's obvious that it's a function and not a tracker. If you want to edit the function you would do it in a similar way to editing a tracker: you select the 3 dots in the top right and go to edit, this brings you back to the code editor activity. 

I think there will come advantages of functions having a location within the group hierarchy and in general I think it's best if basically everything goes in the group hierarchy. So for example I think there would be advantages also for things like reminders being in groups because then when you get a reminder notification and you click on it you get taken to the group for that reminder instead of just the top level group. I think this is a good general direction for the app to take as it allows you to organise things the way you want and provides a lot of power and flexibility whilst also not being imposing for new users. I mean if you just want one page that just contains all your functions you can do it that way too, just create a group called functions and add them all there in the list. To be clear a function should be able to reference data from anywhere, not just its own group.